### PR TITLE
Added from_prior header support

### PR DESCRIPTION
**Pull request by spivachuk**

### CODE REVIEW

**Review by ashcherbakov — APPROVED** [review]
LGTM.
Just a couple of minor comments.

File: didcomm/core/from_prior.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    Args:
+        message: a plaintext message as a dict which optionally contains from_prior not yet packed to JWS
+        resolvers_config: secrets and DIDDoc resolvers
+        issuer_kid: optionally provided issuer key to use for signing from_prior
+
+    Returns:
+        identifier of the issuer key actually used to sign from_prior
+    """
+    if message.get("from_prior") is None:
+        return None
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/core/from_prior.py:35]: Fixed.

File: didcomm/core/from_prior.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    from_prior = message["from_prior"]
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/core/from_prior.py:37]: Added an item to https://github.com/sicpa-dlab/didcomm-gemini/issues/95

File: didcomm/unpack.py [hunk 3]
```diff
@@ -107,6 +108,13 @@ async def unpack(
         metadata.sign_from = unwrap_sign_result.sign_frm_kid
         metadata.sign_alg = unwrap_sign_result.alg
 
+    if msg_as_dict.get("from_prior") is not None:
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/unpack.py:111]: https://github.com/sicpa-dlab/didcomm-gemini/issues/95

File: didcomm/core/from_prior.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    if not isinstance(from_prior, dict):
+        raise MalformedMessageError(MalformedMessageCode.INVALID_PLAINTEXT)
+
+    if from_prior["sub"] == from_prior["iss"]:
+        raise DIDCommValueError()
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/core/from_prior.py:43]: There is a corresponding item in https://github.com/sicpa-dlab/didcomm-gemini/issues/95

File: didcomm/pack_plaintext.py [hunk 5]
```diff
@@ -25,11 +32,52 @@ async def pack_plaintext(resolvers_config: ResolversConfig, message: Message) ->
 
     :param resolvers_config: secrets and DIDDoc resolvers
     :param message: The message to be packed into a DIDComm message
-
+    :param pack_params: Optional parameters for pack
 
     :raises DIDNotResolvedError: If a DID or DID URL (key ID) can not be resolved or not found
     :raises SecretNotResolvedError: If there is no secret for the given DID or DID URL (key ID)
 
     :return: A packed message as a JSON string.
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/pack_plaintext.py]: Fixed.

File: didcomm/pack_signed.py [hunk 6]
```diff
@@ -48,11 +52,24 @@ async def pack_signed(
 
     :return: A packed message as a JSON string.
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/pack_signed.py]: Fixed.

File: didcomm/unpack.py [hunk 7]
```diff
@@ -154,10 +165,12 @@ class Metadata:
     encrypted_from: Optional[DID_URL] = None
     encrypted_to: Optional[List[DID_URL]] = None
     sign_from: Optional[DID_URL] = None
+    from_prior_issuer_kid: Optional[DID_URL] = None
     enc_alg_auth: Optional[AuthCryptAlg] = None
     enc_alg_anon: Optional[AnonCryptAlg] = None
     sign_alg: Optional[SignAlg] = None
     signed_message: Optional[JWS] = None
+    signed_from_prior: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/unpack.py]: Renamed.

File: didcomm/core/from_prior.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    if message.get("from") is not None and from_prior["sub"] != message["from"]:
+        raise DIDCommValueError()
+
+    if issuer_kid is not None and get_did(issuer_kid) != from_prior["iss"]:
+        raise DIDCommValueError()
+
+    issuer_did_or_kid = issuer_kid or from_prior["iss"]
+
+    jws = JsonWebSignature()
```
[the remark this answers is not in the document]
Comment (spivachuk) [reply didcomm/core/from_prior.py]: Done.

**Review by ashcherbakov — APPROVED** [review]

**Pull request merged**